오픈마켓 I [STEP 2] 웡빙, 보리사랑

yusw10 opened the pull request:
# 오픈마켓 STEP2 PR

하리보 이번 STEP2도 잘 부탁드립니다...! 아낌 없는 조언 부탁드립니다🫡
@HARlBO 
@yusw10 
@wongbingg 
###### tags: `PR`

## Step 2 기능 구현 : List와 Grid 스타일의 ModernCollectionView 구현
- MainViewController 를 생성
- SegmentContol 생성
- ModernCollectionView 구현을 위한 준비
<table><tr><td valign="top" width="50%" align="center" border="1">
List View

</td><td valign="top" width="50%" align="center">
Grid View
</td></tr>
<tr><td valign="top" width="50%">

- ListCell  
- createListLayout()
- configureListDataSource()
</td><td valign="top" width="50%">

- GridCell  
- createGridLayout()
- configureListDataSource()
</td></tr>
</table>

- SegmentControl을 이용한 뷰 전환 구현
    - selectecSegmentIndex 프로퍼티를 활용하여 값이 바뀌는 시점을 이용
    - setCollectionViewLayout(_:, animated:) 메서드로 layout을 변경
    - collectionView.dataSource 를 변경
    - collectionView의 Layout을 변경
- 조건별로 달라지는 레이블 설정
    - 재고수량
        - 0일 경우) text: "품절" , textColor: orange
        - 0보다 클 경우) text: "재고수량 : \(재고수량)" , textColor: lightGray
    - 할인가격
        - 할인이 들어갔을 경우) 기존가격의 textColor: red, 기존가격에 취소선을 추가
        - 할인이 들어가지 않은 경우) bargainPrice 를 숨김처리

- 로딩화면에 activity indicator 추가
    - 데이터를 받아오기 시작할 때, 로딩중을 표시할 수 있도록 activity indicator를 시작
    - 데이터를 모두 받아온 뒤, 셀이 로드 되면 activity indicator를 멈춤

![](https://i.imgur.com/kWO2ZLK.gif)


## UML
### Class Diagram(Step2 PR 기준)
![](https://i.imgur.com/K6uByTP.png)



## 궁금한 점 / 고민한 점

### Class Diagram 표현 범위
Step2에서 저희가 구현한 내용의 대부분은 URLSession을 통해 실제로 데이터를 불러오는 과정이었습니다. 특히 `MainViewController`에서 많은 작업을 진행했는데, 이런 경우 Class diagram에서 표현을 모두 해주는지, 아니면 이렇게 UI관련된 리소스들을 표현할 때 사용하시는 UML이 있는지 궁금합니다.

### ListView에서 스크롤 시, 셀 높이 축소현상 이슈
저희가 초기에 List Layout을 구현한 방법은 다음과 같습니다. 
1. List Custom Cell 구현
2. Apple에서 제공하는 기본 List양식을 활용하여 리스트 레이아웃 구성
```swift
func createListLayout() -> UICollectionViewCompositionalLayout {    
    let config = UICollectionLayoutListConfiguration(appearance: .plain)
    return UICollectionViewCompositionalLayout.list(using: config)
}
```
3. 이를 바탕으로 모던 콜렉션뷰 구현

하지만 위와 같은 방식으로 구현을 하니 스크롤을 할 경우 커스텀으로 구현한 리스트 셀의 이미지가 압축되어 표현되었습니다. 이를 해결하기 위해 

1. compression resistance를 극한으로 올려 절대 압축되지 않게 하기
2. Cell 내부의 모든 Heigth를 constant값을 주어 정확하게 표현되게하기
3. 오토 레이아웃으로 높이 안줄어들게 하기

등등 다양한 방법을 시도했었습니다. 다만 어떻게 해도 셀의 높이가 축소되었습니다. 

현재는 기존의 list 레이아웃을 구현하는 것에서 직접 Compositional Layout을 사용하여 리스트형태로 출력되게 만들었는데, 왜 이렇게 출력이 된건지 아직도 이해할 수가 없었습니다. 이 문제에 대해서 어떤 키워드로 검색을 해보면 좋을지 궁금했습니다.

<img src=https://i.imgur.com/4TWv7Yl.png width=50%>

### List 레이아웃에서 Grid 레이아웃으로 넘어갈때 한줄 아래의 컨텐츠를 보여주는 이슈
현재 List 상태에서 Grid 상태로 넘어가게 되면 List 레이아웃의 가장 위쪽 컨텐츠를 보고 있더라고 Grid에서는 3,4번째 컨텐츠(2번쨰 줄)부터 출력이 되는걸 확인했습니다. 이전 계산기 처럼 옮겼을 때 맨위로 옮기는 건 적절치 않다고 생각했습니다. 이를 해결해보고 싶은데 어떤 내용을 찾아보면 좋을지 궁금합니다. (뭘 찾아봐야 하는지 키워드를 모르겠습니다.)

## 트러블 슈팅

### reuse cell 이슈
스크롤 마다 이전 셀에서의 프로퍼티 값이 그대로 남아있어 중복적용되는 현상이 발생했습니다. 재고수량 label 을 예로들면, 품절일 경우 textColor를 orange로 해주었는데, 스크롤을 할 때마다 품절이 아닌 경우에도 textColor가 orange로 변경되어 결국엔 모든 재고수량 label이 orange 색으로 변경 되었습니다. 이를 해결하기 위해 ListCell 클래스 내부에서 override func prepareForReuse() 메서드를 통해 재고수량 label의 textColor 를 nil 로 초기화 해주었습니다.

yusw10 committed:
feat: MainViewController 생성 및 기본 view 선언

- SceneDelegate를 통해 코드로 프로그램이 시작하게 선언
- MainStroyBoard 삭제
- SegmentController 및 교체될 뷰 두개 선언

wongbingg committed:
feat: SegmentControl 추가

- UISegmentContol 인스턴스를 만들어 네비게이션에 적용
- segmentControl의 프로퍼티 selectedSegmentIndex 를 didChangeValue() 메서드에 이용하여 뷰 전환 구현
- firstView 와 secondView의 Layout 설정

yusw10 committed:
refactor: ModernCollectionView 사용하기 위해 코드 정리

- dataFetch 메서드를 데이터 타입 안에 구현
- 기존의 뷰를 바꾸는 방식에서 모던 컬렉션 뷰를 사용하기 위해 불필요한 뷰 제거

yusw10 committed:
feat: Modern Collection View List Layout 구현

- Product 타입이 Hashable을 채택하도록 수정
- Modern Collecion View을 기본 리스트 타입으로 사용하기 위해 Layout과 dataSource 및 셀 생성 구현
- 데이터 통신을 하기 위해 NetworkManager를 프로퍼티로 생성

wongbingg committed:
feat: "List"와 "Grid" 형태의 커스텀셀과 레이아웃 메서드 생성

wongbingg committed:
refactor: segmentControl 을 이용한 뷰전환 구현

- shouldHideListLayout 프로퍼티를 기반으로, 기존의 뷰를 지워준 뒤 Layout 과 DataSource를 재생성, 적용

yusw10 committed:
doc: README.md update

wongbingg committed:
refactor: 각 Cell의 레이블 설정 변경

- 가격 레이블에 Decimal 스타일 적용
- 재고가 없을 경우 잔여수량 레이블을 "품절" 로 표시하도록 조건문 추가

wongbingg committed:
refactor: Data fetch 코드를 메서드로 분리

- manager.dataTask 클로저를 메서드로 분리하여 사용

yusw10 committed:
feat: ListCell에 할인 가격(productBargainLabel)을 추가

- 가격 할인이 된 상품이라면 할인가격과 현재판매가격을 동시에 출력하도록 구현
- productBargainPriceLabel과 productPriceLabel을 가로로 출력하기 위해 horizonal Stack View에 담아서 구현

yusw10 committed:
refactor: 출력되는 데이터와 라벨의 매핑 수정

- 왼쪽에 출력되는 label이 productPriceLabel이 되도록 수정

yusw10 committed:
refactor: Cell을 reuse 하면서 생기는 이슈 해결

- 할인된 가격과 원가가 같을 경우와 다를경우를 분리하여 Cell에 표현하도록 구현
- productPriceLabel과 productBargainPriceLabel이 재사용되면서 label 속성이 불필요하게 재사용 되는 현상 수정

wongbingg committed:
refactor: ListLayout 메서드 수정

- gridLayout 을 구성하는 방식을 가져와 사용
- 스크롤 시, 셀 높이가 줄어드는 오류 해결

yusw10 committed:
refactor: List 와 Grid 스타일 변환시 자연스럽게 변경되도록 수정

- 기존 dataSource를 list와 grid 스타일별로 분리
- Layout도 list와 grid 스타일을 각각 프로퍼티로 저장
- SegmentController에서 각 스타일 및 데이터소스를 교체하도록 selector 메서드 수정

yusw10 committed:
feat: 데이터 로드시 activityIndicator를 표시하도록 구현

- dataTask 메서드를 통해 각 dataSource에 데이터를 로드할 시 indicator가 로딩중임을 표시할 수 있도록 구현

HARlBO commented on OpenMarket/OpenMarket/GridCell.swift:
```diff
@@ -0,0 +1,153 @@
+
+import UIKit
+
+class GridCell: UICollectionViewCell {
```
상속이 되지 않을 class 에는 `final` 키워드 붙이는거 잊지 마세요 😉

wongbingg replied:
이 부분을 신경쓰지 못했군요! 수정 하였습니다 !ㅎㅎ

HARlBO commented on OpenMarket/OpenMarket/GridCell.swift:
```diff
@@ -0,0 +1,153 @@
+
+class GridCell: UICollectionViewCell {
+    
+    let numberFormatter = NumberFormatter()
```
numberFormatter 도 파일내부에서만 사용되는데 `private` 붙여주세요!

wongbingg replied:
numberFormatter 적용 로직을 Int 의 extension 에 메서드로 만들어 주었습니다 !

HARlBO commented on OpenMarket/OpenMarket/GridCell.swift:
```diff
@@ -0,0 +1,153 @@
+    private let productImageView: UIImageView = {
+        let imageView = UIImageView()
+        imageView.translatesAutoresizingMaskIntoConstraints = false
+        imageView.contentMode = .scaleAspectFit
```
contentMode 에는 어떤 것들이 있고 `.scaleAspectFit` 을 선택하신 이유는 어떤걸까요? 🧐

yusw10 replied:
마켓에서 보통 제품사진을 표시할 때 작더라도 전체적인 제품의 형태를 보여주는걸 생각했습니다. 가령 청소기를 넣는다면 저희가 이미지뷰의 비율을 1:1로 만들었기 때문에 `scaleAspectFill`을 사용하게 되면 잘려서 보여진다고 생각했습니다.

그래서 비율을 유지하면서 뷰의 사이즈에 맞게 이미지를 늘리는 방식인 `scaleAspectFit`을 선택하여 사용하였습니다.

HARlBO commented on OpenMarket/OpenMarket/GridCell.swift:
```diff
@@ -0,0 +1,153 @@
+        self.layer.cornerRadius = 12
+    }
+    
+    func setupCellData(with inputData: Product) {
```
setupCellData, inputData  data 가 중복으로 들어가고
함수 호출시에 cell.setupCellData(..) cell 이 중복이 되어서
함수명 자체는 `setup` 정도로해도 좋을것 같아요 👀

yusw10 replied:
호출부분에서 말씀하신대로 쭉 읽어보니 cell과 data가 두번씩 읽혀서 가독성이 떨어지는걸 확인했습니다.
제안해주신 대로 `setup`으로 rename 시켜보았습니다.

HARlBO commented on OpenMarket/OpenMarket/GridCell.swift:
```diff
@@ -0,0 +1,153 @@
+        }
+    }
+    
+    override func prepareForReuse() {
```
함수를 선언할때도 순서를 맞춰주는게 좋아요!
@IBAction, 변수 선언, 뷰의 라이프 사이클, 함수, @IBOutlet, @objc
꼭 이 순서가 아니더라도 협업할 때 미리 이런식으로 순서를 정해준다면 
다른 사람의 코드를 볼 때 어디쯤엔 이 코드가 있겠구나 라고 예측할 수 있습니다!

변수, 함수도 주로 private -> public 순으로 선언해주면 좋습니다 🪄

yusw10 replied:
초반에 웡빙이랑 이야기 할 때 선언 순서를 호출하는 흐름대로 작성하자고 이야기를 했었습니다. 
가령 현재 `MainViewController`를 보면 
```swift 
addUIComponents()
setupSegment()
configureListDataSource()
configureGridDataSource()
configureHierarchy( )
```
이러한 흐름대로 호출이 되고 있으며 메서드도 이 순서대로 구현했습니다. 
또 @objc 와 같은 중첩 메서드의 경우 사용되는 지점 바로 아래에 구현을 해보았습니다. 

위에서 아래로 코드를 타고 내려간다는 느낌에 맞추기 위해 이렇게 구현하되 이 흐름을 지키는 선에서는 private을 먼저 선언하려고 했습니다. 
예를 들어 `setupSegment` 메서드에서 selector에 들어가는 `didChangeValue` 메서드가 코드 최하단에 위치하면 가독성이 좋지 않을것이라고 생각했습니다. 

결론은, 웡빙과 저는 사전에 코드의 흐름대로 선언을 하는 방식으로 구현하기로 했었는데 이방식은 좋지 않은 습관일까요?

HARlBO replied:
사실 정답이 있는게 아니라서, 두분이 미리 상의해서 정해두신 순서대로 사용하시면 됩니다 👍🏻

HARlBO commented on OpenMarket/OpenMarket/ListCell.swift:
```diff
@@ -0,0 +1,216 @@
+
+import UIKit
+
+class ListCell: UICollectionViewCell {
```
요기도 `final` 잊지 말아주세요 👀

wongbingg replied:
이 부분도 마찬가지로 수정 했습니다! 👍👍

HARlBO commented on OpenMarket/OpenMarket/ListCell.swift:
```diff
@@ -0,0 +1,216 @@
+
+class ListCell: UICollectionViewCell {
+    
+    let numberFormatter = NumberFormatter()
```
`numberFormatter` 가 GridCell과 중복으로 사용되고 있는데 extension으로 빼주는건 어떨까요?

wongbingg replied:
말씀대로 extension으로 분리 해주었습니다 ! ☺️☺️

HARlBO commented on OpenMarket/OpenMarket/ListCell.swift:
```diff
@@ -0,0 +1,216 @@
+        self.productImageView.loadImage(url: url)
+        self.productNameLabel.text = inputData.name
+//        let price = numberFormatter.string(from: NSNumber.init(value: inputData.price))
+//        self.productPriceLabel.text = "\(inputData.currency) " + (price ?? "")
```
더이상 사용되지 않는 코드 주석들은 다 제거해주세요 !

wongbingg replied:
앗 죄송합니다..! 빨리 PR을 보내야겠다는 생각에.. 이부분을 보지 못했군요ㅜㅜ 좀 더 신경 쓰도록 하겠습니다 🧐🧐

HARlBO commented on OpenMarket/OpenMarket/MainViewController.swift:
```diff
@@ -0,0 +1,205 @@
+    }()
+    
+    
+    private var collectionView = UICollectionView(frame: .zero, collectionViewLayout: UICollectionViewLayout())
```
가장 상단에 있는 `shouldHideListLayout` 에서 collectonView 를 사용하고 있어서
collectionView 에 대한 선언이 가장 상단에 있으면 코드 읽기가 더 수월할 것 같아요!

yusw10 replied:
UIComponents 끼리(두개이긴 하지만...) 뭉쳐있는게 더 이해에 도움이 될거라 생각해서 collectionView의 선언을 SegmentController와 함께 두었었는데 이부분도 최상단으로 변경해주었습니다..!

HARlBO commented on OpenMarket/OpenMarket/MainViewController.swift:
```diff
@@ -0,0 +1,205 @@
+        self.listLayout = createListLayout()
+        self.gridLayout = createGridLayout()
+        addUIComponents()
+        activitiIndicator.startAnimating()
```
`startAnimating()` 가 중복으로 들어가 있어요 👀

yusw10 replied:
main.async를 사용해서 indicator를 관리할 때 넣은 코드인데 중복인지 확인을 못했던거 같습니다.
이부분도 한쪽 제거했습니다!

HARlBO commented on OpenMarket/OpenMarket/MainViewController.swift:
```diff
@@ -0,0 +1,205 @@
+    
+    private func addUIComponents() {
+        self.navigationItem.titleView = segmentController
+        self.view.addSubview(activitiIndicator)
```
여기서 activityIndicator 를 add 해주면
collectionView 보다 먼저 add 가 되어서 네비게이션바가 의도대로 작동되지 않고 있어요!
확인해보시고 add 되는 위치를 수정해주세요 😉

yusw10 replied:
어... 질문의 의미를 정확하게 이해한건지 잘 모르겠습니다.
인디케이터가 collectionView보다 먼저 등록이 되어서 인디케이터가 콜렉션 뷰에 가려지기 떄문에 안보일 수 있다는 말씀이신가요??
fetch 가 되는 시점에서 collectionView의 alpha값을 0과 1로 수정하도록 구현해서 저희는 잘 나온다고 생각을 했습니다.
다만 alpha를 건드리지 않아도 progress 창이 나오도록 addSubview 하는 순서를 CollectionView가 먼저 등록되도록 수정 해보았습니다.

HARlBO replied:
전 코드 빌드를 돌려보면 segementControl 이 있는 네비게이션이 투명하고, 영역이 잡히지 않았는데요
요번에 수정하신거 보면 네비게이션 영역이 잡혀있는걸 볼 수 있어요! 그 부분 봐주시면 될것 같아요 😉

HARlBO commented on OpenMarket/OpenMarket/MainViewController.swift:
```diff
@@ -0,0 +1,205 @@
+    private func addUIComponents() {
+        self.navigationItem.titleView = segmentController
+        self.view.addSubview(activitiIndicator)
+        self.navigationItem.rightBarButtonItem = UIBarButtonItem(barButtonSystemItem: .add, target: self, action: #selector(addButtonDidTapped))
```
다른곳에서도 많이 보이는데
self.navigationItem 에 대한 세팅들 사이에 self.view.addSubview() 가 들어가 있어요
같은 속성들에 대한 세팅들은 모아주세요 👀

yusw10 replied:
앗... 기능 구현을 하면서 한줄한줄 넣어주다보니 이렇게 된 것 같습니다...
view와 navigationItem 각각의 속성끼리 묶여있도로 수정해보았습니다.

HARlBO commented on OpenMarket/OpenMarket/MainViewController.swift:
```diff
@@ -0,0 +1,205 @@
+}
+// MARK: - Modern Collection VIew Configure Datasource
+extension MainViewController {
+    private func configureHierarchy(){
```
사소한거지만 () { 사이에 공백 추가해주세요 😉

yusw10 replied:
컨벤션 수정했습니다..!

HARlBO reviewed (commented):
안녕하세요 웡빙, 보리사랑 🙋🏻‍♀️
STEP2 도 고생많으셨어요! 코드에서 고민과 고생의 흔적들이 보였답니다 👏🏻

### Class Diagram 표현 범위
ViewController 의 경우 UIViewController 상속부터 view life cycle 관련 함수, 구현한 함수들, delegate 들 등 UML 에서 표현해야 할 요소들이 너무 많기 때문에 모두 포함시키기는 힘들 것 같아요!
그래서 MainViewController 내부 프로퍼티들 보다는 다른 클래스들과의 관계정도 표현해주시면 될 것 같습니다!

### ListView에서 스크롤 시, 셀 높이 축소현상 이슈
저는 요 현상이 재현이 안되는데 호옥시 수정하신게 있을까요?
보통 저런 이슈는 오토레이아웃이 잘못 걸린 케이스가 많아요 🥲 재현 되실때 오토레이아웃 부분 코드 첨부해주시면 같이 보면 좋을것 같아요!

### List 레이아웃에서 Grid 레이아웃으로 넘어갈때 한줄 아래의 컨텐츠를 보여주는 이슈
CollectionView 에 `scrollToItemAtIndexPath` 이라는 메소드가 있어요!
한번 적용해보시고 어려운점이 있다면 말씀 주세요 😉

---

이번에 코드를 보면 전체적으로 코드가 정리되지 않은 느낌이 사알짝 있었어요 🙊
코멘트 드렸던 부분들이 해당 라인에만 적용되는게 아니라 전체적으로 한번씩 보고 수정해주세요!
같은 프로퍼티에 대한 속성들 세팅하는 코드는 라인으로 묶어주시고,
사용하지 않는 주석 제거, 공통 코드들 묶어주기, 함수 선언 순서들 등 
정해진 시간에 학습하고 코드 구현하시느라 힘드시겠지만 아쥬 살짝만 신경써주세요 ☺️

wongbingg committed:
refactor: extension 분리 및 주석제거

- extension 분리 파일을 만들어 폴더로 그룹화
    - NumberFormatter 적용 부분을 extension 으로 분리
    - UIImageView 이미지url을 받아서 이미지를 load하는 로직 분리
    - UILabel에 취소선 추가하는 로직 분리
- 불필요한 주석 제거

wongbingg committed:
refactor: 접근제어자 변경 및 final class 처리

메서드와 프로퍼티 위치 알맞게 변경

yusw10 committed:
refactor: 셀의 데이터 주입 메서드 수정

- GridCell의 setupCellData를 setup으로 수정
- ListCell의 setupCellData를 setup으로 수정

yusw10 committed:
refactor: 메서드 선언 순서 정렬

- 호출 흐름순서대로 메서드를 재배치

yusw10 committed:
refactor: 중복되는 startAnimating 메서드의 제거

- MainViewController 내에서 중복으로 사용되는 startAnimating 메서드 정리

yusw10 committed:
refactor: CollectionView의 등록 시기 수정

- indicator View가 CollectionView 앞쪽으로 올 수 있게 addSubVIew메서드 호출 시기 변경

yusw10 committed:
refactor: 같은 속성을 변경하는 코드끼리 보이도록 선언 순서 변경

yusw10 commented:
전체적으로 Modern CollectionView를 구현하려고 웡빙하고 너무 고민을해서 구현하고 정상 작동하는거 보자마자 얼른 검사 맡고싶어서 정리되지 않은 코드들로 PR 을 보내게 된것같습니다... 다음부터는 읽기 편하시도록 좀 더 가다듬고 올려보도록 하겠습니다...!

그리고 List 에서 축소되는 이슈는 현재는 Compositional Layout으로 구현해서 **현재는 해결된 문제인데 원래 코드에서 정확하게 어떤 원리로 축소가 되는지 궁금해서 여쭤본 사항이었습니다.**
```swift 
private func createLayout() -> UICollectionViewLayout {
    let config = UICollectionLayoutListConfiguration(appearance: .plain)
    return UICollectionViewCompositionalLayout.list(using: config)
}
```
기존에 축소가 되던 리스트의 레이아웃을 애플에서 제공하던 모던 컬렉션뷰 기본 예제를 사용해서 그런것 같긴 한데 셀의 높이를 강제로 주어도 축소가 되는게 이해가 안되었어서 질문드렸었습니다...!

yusw10 committed:
docs: README.md Update

wongbingg committed:
feat: NSCache 를 이용하여 이미지를 캐시하는 메서드 구현

HARlBO reviewed (approved):
Cell 의 이미지가 축소되었던 이슈는 오토레이아웃이 잘못 걸려있어서 발생했을 가능성이 가장 큽니다!

지금도 빌드를 해보면 
<img width="1112" alt="image" src="https://user-images.githubusercontent.com/13211322/180493112-ef16f016-ec13-43d1-a65a-d33113e85647.png">
요런식으로 레이아웃이 깨졌다는 로그가 엄청나게 보이는데요...!
레이아웃을 다시 잡아보시는게 좋을것 같아요.
지금은 중복되거나 누락된  constraint 들을 Xcode 가 임의로 잡아주고 있어서
화면상으로는 문제가 없어 보였지만 사실 레이아웃들이 많이 깨져 있는 상태 입니닷 🥲
아마 이미지 축소되엇던 문제도 레이아웃 이슈였을것 같네요!

오토레이아웃 한번 두분이서 다시 잡아보시고 질문이 생기면 언제든 편하게 질문주세요!
나머지 변경사항은 확인 완료 했습니다 👏🏻 
고생많으셨어요 🙏🏻
